### PR TITLE
add: guide on kerascv conversion tool.

### PR DESCRIPTION
**Pull request by sayakpaul**
As internally [discussed](https://huggingface.slack.com/archives/C03UQJENJTV/p1674886674724979).

### CODE REVIEW

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+* [DreamBooth](https://keras.io/examples/generative/dreambooth/)
+* [Prompt-to-Prompt editing](https://github.com/miguelCalado/prompt-to-prompt-tensorflow)
+
+A Stable Diffusion model has three primary components:
+
+* Text encoder 
+* UNet 
+* VAE 
+
+Depending on the fine-tuning task, we may fine-tune one of more of these components (the VAE is almost always left untouched). Here are some common combinations:
```
Comment (patil-suraj) [inline docs/source/en/using-diffusers/kerascv.mdx:53]:
```suggestion
Depending on the fine-tuning task, we may fine-tune one or more of these components (the VAE is almost always left untouched). Here are some common combinations:
```

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+* DreamBooth: Text encoder and UNet
```
Comment (patil-suraj) [inline docs/source/en/using-diffusers/kerascv.mdx:55]: (nit)
```suggestion
* DreamBooth: UNet and Text encoder 
```

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+* Text encoder 
+* UNet 
+* VAE 
+
+Depending on the fine-tuning task, we may fine-tune one of more of these components (the VAE is almost always left untouched). Here are some common combinations:
+
+* DreamBooth: Text encoder and UNet 
+* Classical text2image: UNet 
+* Textual Inversion: Text encoder
```
Comment (patil-suraj) [inline docs/source/en/using-diffusers/kerascv.mdx:57]:
```suggestion
* Textual Inversion: Just the newly initialized embeddings in the text encoder.
```

**Review by patil-suraj — APPROVED** [review]
Very cool, thanks a lot for adding this!

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+<Tip warning={true}>
+
+This is an experimental feature.
+
+</Tip>
+
+[KerasCV](https://github.com/keras-team/keras-cv/) provides APIs for implementing various computer vision workflows. It
+also provides the Stable Diffusion [v1 and v2](https://github.com/keras-team/keras-cv/blob/master/keras_cv/models/stable_diffusion)
+models. Many practitioners find it easy to fine-tune the Stable Diffusion models shipped by KerasCV. However, as of this writing, KerasCV offers limited support to experiment with Stable Diffusion models. On the other hand,
```
Comment (pcuenca) [inline docs/source/en/using-diffusers/kerascv.mdx:23]:
```suggestion
models. Many practitioners find it easy to fine-tune the Stable Diffusion models shipped by KerasCV. However, as of this writing, KerasCV offers limited support to experiment with Stable Diffusion models for inference and deployment. On the other hand,
```

Maybe clarify we are not talking about research experiments?

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+Diffusers provides tooling dedicated to this purpose (and more), such as different [noise schedulers](https://huggingface.co/docs/diffusers/using-diffusers/schedulers), [flash attention](https://huggingface.co/docs/diffusers/optimization/xformers), and [other 
+optimization techniques](https://huggingface.co/docs/diffusers/optimization/fp16).
+
+How about fine-tuning Stable Diffusion models in KerasCV and export them such that they become compatible in Diffusers to combine the
+best of both worlds? We have created a [tool](https://huggingface.co/spaces/sayakpaul/convert-kerascv-sd-diffusers) that
```
Comment (pcuenca) [inline docs/source/en/using-diffusers/kerascv.mdx:28]:
```suggestion
How about combining the best of both worlds by fine-tuning Stable Diffusion models in KerasCV and exporting them to Diffusers? We have created a [tool](https://huggingface.co/spaces/sayakpaul/convert-kerascv-sd-diffusers) that
```

Slight simplification

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+lets you do just that! It takes a KerasCV Stable Diffusion checkpoints and exports them to Diffusers-compatible checkpoints.
```
Comment (pcuenca) [inline docs/source/en/using-diffusers/kerascv.mdx:29]:
```suggestion
lets you do just that! It takes KerasCV Stable Diffusion checkpoints and exports them to Diffusers-compatible checkpoints.
```

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+[`StableDiffusionPipeline`](https://huggingface.co/docs/diffusers/api/pipelines/stable_diffusion/overview) which is ready
+for inference. Finally, it pushes the converted checkpoints to a repository on the Hugging Face Hub. 
+
+We welcome you to try out the tool [here](https://huggingface.co/spaces/sayakpaul/convert-kerascv-sd-diffusers)
+and share feedback via [discussions](https://huggingface.co/spaces/sayakpaul/convert-kerascv-sd-diffusers/discussions/new). 
+
+## Getting started 
+
+First, you need to obtain the fine-tuned KerasCV Stable Diffusion checkpoints. We provide an
+overview of the different ways Stable Diffusion models can be fine-tuned [here](https://huggingface.co/docs/diffusers/training/overview). For Keras implementations of some of these methods, you can check out the resources below:
```
Comment (pcuenca) [inline docs/source/en/using-diffusers/kerascv.mdx:40]:
```suggestion
overview of the different ways Stable Diffusion models can be fine-tuned [using `diffusers`](https://huggingface.co/docs/diffusers/training/overview). For the Keras implementation of some of these methods, you can check out these resources:
```

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+A Stable Diffusion model has three primary components:
+
+* Text encoder 
+* UNet 
+* VAE 
+
+Depending on the fine-tuning task, we may fine-tune one of more of these components (the VAE is almost always left untouched). Here are some common combinations:
+
+* DreamBooth: Text encoder and UNet 
+* Classical text2image: UNet
```
Comment (pcuenca) [inline docs/source/en/using-diffusers/kerascv.mdx:56]:
```suggestion
* Classical text to image fine-tuning: UNet 
```

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+* Text encoder 
+* UNet 
+* VAE 
+
+Depending on the fine-tuning task, we may fine-tune one of more of these components (the VAE is almost always left untouched). Here are some common combinations:
+
+* DreamBooth: Text encoder and UNet 
+* Classical text2image: UNet 
+* Textual Inversion: Text encoder 
+
```
Comment (pcuenca) [inline docs/source/en/using-diffusers/kerascv.mdx:58]: Maybe add another section here?

```suggestion

## Performing the Conversion

```

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+by conducting Textual Inversion with the following "placeholder token": `<my-funny-cat-token>`. 
+
+On the tool, we supply the following things: 
+
+* Path(s) to download the fine-tuned checkpoint(s) (KerasCV)
+* An HF token 
+* Placeholder token (only applicable for Textual Inversion)
+
+<div align="center">
+<img src="https://i.imgur.com/JMATVKJ.png"/>
```
Comment (pcuenca) [inline docs/source/en/using-diffusers/kerascv.mdx:69]: Are we keeping this in `imgur` or moving to documentation images? Same with the others.

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+</Tip>
+
+[KerasCV](https://github.com/keras-team/keras-cv/) provides APIs for implementing various computer vision workflows. It
+also provides the Stable Diffusion [v1 and v2](https://github.com/keras-team/keras-cv/blob/master/keras_cv/models/stable_diffusion)
+models. Many practitioners find it easy to fine-tune the Stable Diffusion models shipped by KerasCV. However, as of this writing, KerasCV offers limited support to experiment with Stable Diffusion models. On the other hand,
+Diffusers provides tooling dedicated to this purpose (and more), such as different [noise schedulers](https://huggingface.co/docs/diffusers/using-diffusers/schedulers), [flash attention](https://huggingface.co/docs/diffusers/optimization/xformers), and [other 
+optimization techniques](https://huggingface.co/docs/diffusers/optimization/fp16).
+
+How about fine-tuning Stable Diffusion models in KerasCV and export them such that they become compatible in Diffusers to combine the
```
Comment (kashif) [inline docs/source/en/using-diffusers/kerascv.mdx:27]:
```suggestion
How about fine-tuning Stable Diffusion models in KerasCV and exporting them such that they become compatible with Diffusers to combine the
```

File: docs/source/en/using-diffusers/kerascv.mdx (hunk 6 above)
Comment (kashif) [inline docs/source/en/using-diffusers/kerascv.mdx:29]:
```suggestion
lets you do just that! It takes a KerasCV Stable Diffusion checkpoints and exports it to Diffusers-compatible checkpoints.
```

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+</div> 
+
+If you head over to the [model card of the repository](https://huggingface.co/sayakpaul/textual-inversion-cat-kerascv_sd_diffusers_pipeline), the
+following should appear: 
+
+<div align="center">
+<img src="https://i.imgur.com/PxCNRFh.png"/>
+</div> 
+
+_**Note that we're not specifying the UNet weights here since the UNet is not fine-tuned during Textual Inversion.**_
```
Comment (pcuenca) [inline docs/source/en/using-diffusers/kerascv.mdx:92]: I don't understand what this means, sorry. Is this about the first step in the conversion process? If so, I'd put this comment (maybe as a `tip`) below the corresponding screenshot.

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+
+And that's it! You now have your fine-tuned KerasCV Stable Diffusion model in Diffusers 🧨
+
+## Usage in Diffusers
```
Comment (pcuenca) [inline docs/source/en/using-diffusers/kerascv.mdx:96]:
```suggestion
## Using the Converted Model in Diffusers 
```

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+<img src="https://i.imgur.com/PxCNRFh.png"/>
+</div> 
+
+_**Note that we're not specifying the UNet weights here since the UNet is not fine-tuned during Textual Inversion.**_
+
+And that's it! You now have your fine-tuned KerasCV Stable Diffusion model in Diffusers 🧨
+
+## Usage in Diffusers 
+
+Just besides the model card of the [repository](https://huggingface.co/sayakpaul/textual-inversion-cat-kerascv_sd_diffusers_pipeline),
```
Comment (kashif) [inline docs/source/en/using-diffusers/kerascv.mdx:98]:
```suggestion
Just beside the model card of the [repository](https://huggingface.co/sayakpaul/textual-inversion-cat-kerascv_sd_diffusers_pipeline), 
```

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+placeholder_token = "<my-funny-cat-token>"
+prompt = f"two {placeholder_token} getting married, photorealistic, high quality"
+image = pipeline(prompt, num_inference_steps=50).images[0]
+```
+
+<div align="center">
+<img src="https://i.imgur.com/fYHgzSc.png"/>
+</div> 
+
+One can also continue fine-tuning from these Diffusers checkpoint leveraging efficient optimization tools from Diffusers. Refer [here](https://huggingface.co/docs/diffusers/main/en/optimization/fp16) for
```
Comment (kashif) [inline docs/source/en/using-diffusers/kerascv.mdx:166]:
```suggestion
One can also continue fine-tuning from these Diffusers checkpoints by leveraging efficient optimization tools from Diffusers. Refer [here](https://huggingface.co/docs/diffusers/main/en/optimization/fp16) for 
```

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+
+On the top right hand side, we provide a "Use in Diffusers" button. If you click the button, you should see the following code-snippet: 
+
+```py 
+from diffusers import DiffusionPipeline
+
+pipeline = DiffusionPipeline.from_pretrained("sayakpaul/textual-inversion-cat-kerascv_sd_diffusers_pipeline")
+```
+
+Let's perform inference!
```
Comment (pcuenca) [inline docs/source/en/using-diffusers/kerascv.mdx:113]:
```suggestion
The model is in standard `diffusers` format. Let's perform inference!
```

(Just to stress it, so they don't need to study the following code in depth if they are already familiar with the API).

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+
+_**Note that if you specified a `placeholder_token` while performing the conversion, the tool will log it accordingly. Refer
+to the model card of [this repository](https://huggingface.co/sayakpaul/textual-inversion-cat-kerascv_sd_diffusers_pipeline)
+as an example.**_
+
+We welcome you to use the tool for various Stable Diffusion fine-tuning scenarios and let us know your feedback! Here are some examples 
+of Diffusers checkpoints that were obtained using the tool: 
+
+* [sayakpaul/text-unet-dogs-kerascv_sd_diffusers_pipeline](https://huggingface.co/sayakpaul/text-unet-dogs-kerascv_sd_diffusers_pipeline)
+* [sayakpaul/unet-dogs-kerascv_sd_diffusers_pipeline](https://huggingface.co/sayakpaul/unet-dogs-kerascv_sd_diffusers_pipeline)
```
Comment (pcuenca) [inline docs/source/en/using-diffusers/kerascv.mdx:140]: Maybe mention where did these come from: dreambooth, fine-tuning?

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+best of both worlds? We have created a [tool](https://huggingface.co/spaces/sayakpaul/convert-kerascv-sd-diffusers) that
+lets you do just that! It takes a KerasCV Stable Diffusion checkpoints and exports them to Diffusers-compatible checkpoints.
+More specifically, it first converts the checkpoints to PyTorch and then wraps them into a
+[`StableDiffusionPipeline`](https://huggingface.co/docs/diffusers/api/pipelines/stable_diffusion/overview) which is ready
+for inference. Finally, it pushes the converted checkpoints to a repository on the Hugging Face Hub. 
+
+We welcome you to try out the tool [here](https://huggingface.co/spaces/sayakpaul/convert-kerascv-sd-diffusers)
+and share feedback via [discussions](https://huggingface.co/spaces/sayakpaul/convert-kerascv-sd-diffusers/discussions/new). 
+
+## Getting started
```
Comment (pcuenca) [inline docs/source/en/using-diffusers/kerascv.mdx:37]:
```suggestion
## Getting Started 
```

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+to the model card of [this repository](https://huggingface.co/sayakpaul/textual-inversion-cat-kerascv_sd_diffusers_pipeline)
+as an example.**_
+
+We welcome you to use the tool for various Stable Diffusion fine-tuning scenarios and let us know your feedback! Here are some examples 
+of Diffusers checkpoints that were obtained using the tool: 
+
+* [sayakpaul/text-unet-dogs-kerascv_sd_diffusers_pipeline](https://huggingface.co/sayakpaul/text-unet-dogs-kerascv_sd_diffusers_pipeline)
+* [sayakpaul/unet-dogs-kerascv_sd_diffusers_pipeline](https://huggingface.co/sayakpaul/unet-dogs-kerascv_sd_diffusers_pipeline)
+
+## Incorporating Diffusers goodies 🎁
```
Comment (pcuenca) [inline docs/source/en/using-diffusers/kerascv.mdx:142]:
```suggestion
## Incorporating Diffusers Goodies 🎁
```

File: docs/source/en/using-diffusers/kerascv.mdx (hunk 15 above)
[the remark this answers is not in the document]
Comment (pcuenca) [reply docs/source/en/using-diffusers/kerascv.mdx:166]: I think that page was meant for inference, not training. Maybe we want to link to the training section too?
Question: do we handle ema/non-ema weights somehow (if they were present in the original models)?

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+```
+
+<div align="center">
+<img src="https://i.imgur.com/fYHgzSc.png"/>
+</div> 
+
+One can also continue fine-tuning from these Diffusers checkpoint leveraging efficient optimization tools from Diffusers. Refer [here](https://huggingface.co/docs/diffusers/main/en/optimization/fp16) for 
+more details.  
+
+## Known limitations
```
Comment (pcuenca) [inline docs/source/en/using-diffusers/kerascv.mdx:169]:
```suggestion
## Known Limitations 
```

File: docs/source/en/_toctree.yml [hunk 21]
```diff
@@ -18,6 +18,8 @@
       title: Configuring Pipelines, Models, and Schedulers
     - local: using-diffusers/custom_pipeline_overview
       title: Loading and Adding Custom Pipelines
+    - local: using-diffusers/kerascv
```
Comment (pcuenca) [inline docs/source/en/_toctree.yml:21]: Unsure about this placement. On the one hand, it's related to "loading" weights, but on the other hand is a special purpose thing that requires a previous training.

**Review by pcuenca — COMMENTED** [review]
Very cool!

File: docs/source/en/using-diffusers/kerascv.mdx (hunk 10 above)
Comment (sayakpaul) [reply docs/source/en/using-diffusers/kerascv.mdx:69]: I will move to `documentation-images` :)

File: docs/source/en/using-diffusers/kerascv.mdx (hunk 15 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply docs/source/en/using-diffusers/kerascv.mdx:166]: Sure, any link you have in mind? 

> Question: do we handle ema/non-ema weights somehow (if they were present in the original models)?

If EMA is used, it's just returned as is such that it behaves similarly in TensorFlow / Keras. So, on the conversion side, it won't matter.

File: docs/source/en/using-diffusers/kerascv.mdx (hunk 4 above)
Comment (sayakpaul) [reply docs/source/en/using-diffusers/kerascv.mdx:23]: Nothing salient comes to mind for commenting about it.

File: docs/source/en/using-diffusers/kerascv.mdx (hunk 15 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply docs/source/en/using-diffusers/kerascv.mdx]: In my Keras implementations, if you use EMA during fine-tuning, the EMA'd model is returned such that it won't need any separate treatment during model loading. So, the EMA status of a model doesn't affect the conversion workflow here. 

If you used EMA, only the EMA'd checkpoints are returned in my Keras implementations. But if this behavior is customized, then it's up to the user to specify the weights accordingly.

**Comment by sayakpaul** [comment]
@kashif @patil-suraj @pcuenca thanks a lot for your reviews!

I have addressed all your comments.

File: docs/source/en/using-diffusers/kerascv.mdx [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+
+First, you need to obtain the fine-tuned KerasCV Stable Diffusion checkpoints. We provide an
+overview of the different ways Stable Diffusion models can be fine-tuned [using `diffusers`](https://huggingface.co/docs/diffusers/training/overview). For the Keras implementation of some of these methods, you can check out these resources:
+
+* [Teach StableDiffusion new concepts via Textual Inversion](https://keras.io/examples/generative/fine_tune_via_textual_inversion/)
+* [Fine-tuning Stable Diffusion](https://keras.io/examples/generative/finetune_stable_diffusion/)
+* [DreamBooth](https://keras.io/examples/generative/dreambooth/)
+* [Prompt-to-Prompt editing](https://github.com/miguelCalado/prompt-to-prompt-tensorflow)
+
+A Stable Diffusion model has three primary components:
```
Comment (patrickvonplaten) [inline docs/source/en/using-diffusers/kerascv.mdx:47]: (nit) when we refer to `components` in `diffusers` we usually include the `scheduler` as well. 
See, e.g.:
- https://huggingface.co/docs/diffusers/api/pipelines/overview#pipelines
- https://github.com/huggingface/diffusers/blob/180841bbde4b200be43350164eef80c93a68983a/src/diffusers/pipelines/pipeline_utils.py#L775

both include the scheduler as components. Would rather use "models" here

**Comment by sayakpaul** [comment]
> Very cool! Can you merge yourself or should I merge?

Thanks! I will address the comments first and then will merge. Just wanted to confirm if you are okay with the current position of the doc.

**Comment by sayakpaul** [comment]
I guess the failing test is unrelated to the PR? If so, feel free to merge it as I believe it's good to be merged. 

@patil-suraj @patrickvonplaten @pcuenca 

(I can also merge but didn't want to merge with a failing test mistakenly)

**Comment by patil-suraj** [comment]
yeah the failing test is unrelated , merging the PR!

**Pull request merged**